### PR TITLE
zigbee: osif: Add variant for incomplete platforms

### PR DESCRIPTION
**Pull request by tomchy**
Introduce a simplified variant of OSIF layer parts, to reduce the dependencies between ZBOSS and peripheral drivers.
This variant is not meant do be used in production. It can be useful during:
 - Porting Zigbee support to a new, incomplete platform.
 - Writing component tests (i.e. NVRAM migration or loading)

Signed-off-by: Tomasz Chyrowicz <tomasz.chyrowicz@nordicsemi.no>

### CODE REVIEW

**Comment by tomchy** [comment]
The following symbols will be added before leaving draft stage:
```
	imply COUNTER
	imply COUNTER_TIMER2
```
They are temporary removed to test the ktimer implementation.

**Comment by thst-nordic** [comment]
This PR needs to be rebased, and then the test-spec will be generated.
The CI-* Github Labels are now redundant so I will remove them.

File: doc/nrf/releases/release-notes-changelog.rst [hunk 1]
```diff
@@ -65,6 +65,14 @@ Thread
 Zigbee
 ------
 
+* Added:
+
+  * Zigbee timer implementation, based on the Zephyr system time.
+    This timer implementation is enabled by default in NCP sample and while using ZBOSS development libraries.
```
Comment (greg-fer) [inline doc/nrf/releases/release-notes-changelog.rst:71]:
```suggestion
    This timer implementation is enabled by default in the :ref:`Zigbee NCP sample <zigbee_ncp_sample>` and when using ZBOSS development libraries.
```

File: doc/nrf/releases/release-notes-changelog.rst [hunk 2]
```diff
@@ -65,6 +65,14 @@ Thread
...
+  * Kconfig option to set the number of ZBOSS NVRAM virtual pages.
+  * RAM-based ZBOSS NVRAM implementation.
+    This implementation must not be used in production. It is meant to be used only for debugging purposes.
```
Comment (greg-fer) [inline doc/nrf/releases/release-notes-changelog.rst:74]:
```suggestion
    This implementation must not be used in production.
    It is meant to be used only for debugging purposes.
```

File: subsys/zigbee/Kconfig [hunk 3]
```diff
@@ -363,6 +361,32 @@ config ZIGBEE_USE_BUTTONS
 	imply GPIO
 	imply DK_LIBRARY
 
+config ZIGBEE_NVRAM_PAGE_COUNT
+	int "The number of ZBOSS (NV)RAM pages"
```
Comment (greg-fer) [inline subsys/zigbee/Kconfig:365]:
```suggestion
	int "The number of ZBOSS RAM or NVRAM pages"
```
Global edit

File: doc/nrf/releases/release-notes-changelog.rst [hunk 4]
```diff
@@ -65,6 +65,14 @@ Thread
 Zigbee
 ------
 
+* Added:
+
+  * Zigbee timer implementation, based on the Zephyr system time.
+    This timer implementation is enabled by default in NCP sample and while using ZBOSS development libraries.
+  * Kconfig option to set the number of ZBOSS NVRAM virtual pages.
```
Comment (greg-fer) [inline doc/nrf/releases/release-notes-changelog.rst:72]:
```suggestion
  * :kconfig:option:`CONFIG_ZIGBEE_NVRAM_PAGE_COUNT` Kconfig option to set the number of ZBOSS NVRAM virtual pages.
```

**Review by greg-fer — CHANGES_REQUESTED** [review]

File: doc/nrf/libraries/zigbee/osif.rst [hunk 5]
```diff
@@ -41,6 +41,10 @@ You can also configure the following OSIF-related Kconfig options:
 * :kconfig:option:`CONFIG_ZIGBEE_USE_LEDS` - LEDs abstract for the ZBOSS OSIF layer.
   You can use this option if you want to test ZBOSS examples directly in the |NCS|.
 * :kconfig:option:`CONFIG_ZIGBEE_USE_SOFTWARE_AES` - Configures the ZBOSS OSIF layer to use the software encryption.
+* :kconfig:option:`CONFIG_ZIGBEE_NVRAM_PAGE_COUNT` - Configures the number of ZBOSS (NV)RAM logical pages.
```
[the remark this answers is not in the document]
Comment (tomchy) [reply doc/nrf/libraries/zigbee/osif.rst:44]: If it is unacceptable to use the brackets, I prefer to stay with the `NVRAM` name.

File: subsys/zigbee/osif/zb_nrf_nvram_none.c [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+ */
+
+#include "zboss_api.h"
+
+
+#ifdef ZB_USE_NVRAM
+static uint8_t zb_nvram_buf[CONFIG_ZIGBEE_NVRAM_PAGE_COUNT][CONFIG_ZIGBEE_NVRAM_PAGE_SIZE];
+
+
+/* ZBOSS collaout that should be called once flash erase page operation is finished. */
```
[the remark this answers is not in the document]
Comment (tomchy) [reply subsys/zigbee/osif/zb_nrf_nvram_none.c:14]: fixed

File: subsys/zigbee/osif/zb_nrf_nvram_none.c [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+zb_ret_t zb_osif_nvram_read(zb_uint8_t page, zb_uint32_t pos, zb_uint8_t *buf,
+			    zb_uint16_t len)
+{
+	if ((page < zb_get_nvram_page_count()) &&
+	    ((pos + len) < zb_get_nvram_page_length())) {
+		memcpy(buf, &zb_nvram_buf[page][pos], len);
+	} else {
+		memset(buf, 0xFF, len);
```
[the remark this answers is not in the document]
Comment (tomchy) [reply subsys/zigbee/osif/zb_nrf_nvram_none.c]: I've added error codes instead.

File: doc/nrf/libraries/zigbee/osif.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (greg-fer) [reply doc/nrf/libraries/zigbee/osif.rst]: Yes, ok, just don't use brackets.

**Review by greg-fer — APPROVED** [review]

**Review by sebastiandraus — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Comment by mbolivar-nordic** [comment]
I  believe the compliance failures can be addressed by modifying scripts/checkpatch/typedefsfile.

**Review by carlescufi — APPROVED** [review]

**Pull request merged**